### PR TITLE
Add replacement methods for working with ReferencePositions over LocalReferences

### PR DESCRIPTION
**Pull request by anthony-murphy**
### LocalReference class and method deprecations
The class LocalReference in the @fluidframework/merge-tree packing is being deprecated. Please transition usage to the ReferencePosition interface from the same package.
To support this change the following methods are deprecated with replacements that operate on ReferencePosition rather than LocalReference
 - createPositionReference to createLocalReferencePosition
 - addLocalReference to createLocalReferencePosition
 - localRefToPos to localReferencePositionToPosition
 - removeLocalReference to removeLocalReferencePosition

 The above methods are changes in both the @fluidframework/merge-tree and @fluidframework/sequence packages.

follow up to #10167 and related to #1009

[AB#145](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/145) [AB#148](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/148)

### CODE REVIEW

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 1]
```diff
@@ -2442,6 +2444,31 @@ export class MergeTree {
...
+            if (removedRefs !== undefined) {
+                this.blockUpdatePathLengths(segment.parent, TreeMaintenanceSequenceNumber,
+                    LocalClientId);
+            }
+            return removedRefs;
+        }
+    }
+    public createLocalReferencePosition(
+        segment: ISegment, offset: number, refType: ReferenceType, properties: PropertySet | undefined,
+        client: Client,
```
Comment (ransomr) [inline packages/dds/merge-tree/src/mergeTree.ts:2460]: This is adding the first reference to Client from MergeTree. I had assumed that was intentional - that MergeTree should not be aware of Client. Would it be cleaner to only have the creation API on client and make this an internal method called only by client that takes a LocalReference?

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 2]
```diff
@@ -2442,6 +2444,31 @@ export class MergeTree {
...
+    ): ReferencePosition {
+        const localRefs = segment.localRefs ?? new LocalReferenceCollection(segment);
+        segment.localRefs = localRefs;
+
+        const segRef = localRefs.createLocalRef(offset, refType, properties, client);
+
+        this.blockUpdatePathLengths(segment.parent, TreeMaintenanceSequenceNumber,
```
Comment (ransomr) [inline packages/dds/merge-tree/src/mergeTree.ts:2467]: Why are the calls to blockUpdatePathLengths required? I wouldn't expect creating/removing a reference to change the path lengths.

File: packages/dds/merge-tree/src/referencePositions.ts [hunk 3]
```diff
@@ -11,14 +11,18 @@ import { PropertySet, MapLike } from "./properties";
 export const reservedTileLabelsKey = "referenceTileLabels";
 export const reservedRangeLabelsKey = "referenceRangeLabels";
 
-export const refGetTileLabels = (refPos: ReferencePosition): string[] | undefined =>
+export function refTypeIncludesFlag(refPos: ReferencePosition, flags: ReferenceType): boolean {
```
Comment (ransomr) [inline packages/dds/merge-tree/src/referencePositions.ts:14]: Thank you for adding this - I was getting tired of disabling the lint warning every time I wanted to check if a reference was SlideOnRemove

File: packages/dds/merge-tree/src/localReference.ts [hunk 4]
```diff
@@ -366,22 +424,27 @@ export class LocalReferenceCollection {
...
             for (const lref of iterable) {
-                // eslint-disable-next-line no-bitwise
-                if (lref.refType & ReferenceType.SlideOnRemove) {
-                    beforeRefs.push(lref);
-                    lref.segment = this.segment;
-                    lref.offset = 0;
-                    if (lref.hasRangeLabels() || lref.hasTileLabels()) {
-                        this.hierRefCount++;
+                if (lref instanceof LocalReference) {
+                    if (refTypeIncludesFlag(lref, ReferenceType.SlideOnRemove)) {
```
Comment (ransomr) [inline packages/dds/merge-tree/src/localReference.ts:437]: Is there a way to make the API more clear that SlideOnRemove is only supported on LocalReferences? Would it make sense for setting SlideOnRemove on a marker node be an error?

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 2 above)
Comment (Abe27342) [reply packages/dds/merge-tree/src/mergeTree.ts:2467]: looks to be for bookkeeping on local references in `HierBlock`s, but yeah naming seems confusing

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 1 above)
Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:2460]: this is just staging for back compat. i the next release we will no longer remove client which is deprecated.

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:2467]: the handling for markers vs local references is very confused in the code. i haven't had time to tease it all apart, so for now just going for a do no harm approach where we follow the existing patterns.

File: packages/dds/merge-tree/src/localReference.ts (hunk 4 above)
Comment (anthony-murphy) [reply packages/dds/merge-tree/src/localReference.ts:437]: i want to get there. similar to my above comment there is quite a few places where the code handles either markers or local references, but I don't think it is always necessary. these changes are incrementally cleaning that up

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**